Fix recoverable [U]Int128 division-by-zero

oscbyspro opened the pull request:
This patch fixes the division-by-zero case in the following methods:

- `Int128/dividedReportingOverflow(by:)`
- `Int128/remainderReportingOverflow(dividingBy:)`
- `UInt128/dividedReportingOverflow(by:)`
- `UInt128/remainderReportingOverflow(dividingBy:)`

#### Documentation: /dividedReportingOverflow(by:)

> Dividing by zero is not an error when using this method. For a value x, the result of x.dividedReportingOverflow(by: 0) is (x, true).

#### Documentation: /remainderReportingOverflow(dividingBy:)

> Dividing by zero is not an error when using this method. For a value x, the result of x.remainderReportingOverflow(dividingBy: 0) is (x, true).

oscbyspro committed:
Fix recoverable [U]Int128 division-by-zero

This patch fixes the division-by-zero case in the following methods:

- `Int128/dividedReportingOverflow(by:)`
- `Int128/remainderReportingOverflow(dividingBy:)`
- `UInt128/dividedReportingOverflow(by:)`
- `UInt128/remainderReportingOverflow(dividingBy:)`

oscbyspro commented:
Looking closer at the implementation of `Int128` and `UInt128`, I see that I'll have to add preconditions to `/(a:b:)` and `%(a:b:)` following this change. It also looks like `Int128`'s division operator doesn't trap on ordinary overflow either. I'll mark this pull request as a draft until I have updated `/(a:b:)` and `%(a:b:)`.

oscbyspro committed:
Add preconditions to trapping [U]Int128 division methods.

oscbyspro committed:
Make consistent use of `_slowPath(_:)`.

Jumping on the `_slowPath(_:)` bandwagon like all other integer types.

oscbyspro committed:
Copy existing UInt128 division comments to division operators.

oscbyspro commented:
Legend has it that the cordial ping rate is once a week. @stephentyrone – can you take a look at this pull request? Int128 and UInt128 do not perform division by zero like the smaller binary integers. It would be nice if we could fix that.

stephentyrone commented on stdlib/public/core/Int128.swift:
```diff
@@ -297,8 +297,12 @@ extension Int128 {
   ) -> (partialValue: Self, overflow: Bool) {
-    _precondition(other != .zero, "Division by zero")
-    if self == .min && other == -1 { return (.min, true) }
+    if _slowPath(other == .zero) {
```
Was there an observed performance issue that these `_slowPath` annotations are intended to address, or were they added speculatively?

stephentyrone replied:
More of a matter of my own curiosity. I'm ok with leaving them in, though they shouldn't matter much.

stephentyrone commented on stdlib/public/core/Int128.swift:
```diff
@@ -307,8 +311,12 @@ extension Int128 {
+    if _slowPath(other == .zero) {
+      return (self, true)
+    }
+    if _slowPath(self == .min && other == (-1 as Self)) {
```
This case is somewhat interesting because the remainder _does not_ overflow; rather the analogous division does. It's entirely plausible to me that this was originally defined "wrong", but we're probably stuck with the current behavior. In any case, I think it merits a comment here.

oscbyspro replied:
I return the equivalent of (0, false) in my project. So, I considered asking about this specific interaction, but I figured Swift chose C++ semantics ± trapping on undefined behavior. Is there a particular comment you have in mind?

xwu replied:
This was the specified behavior—I was the one who aligned the actual behavior with the specification in #14219—but for this reason the wording of the corresponding precondition failure (at least at the time) was that "_Division_ results in overflow in remainder operation"

xwu commented on stdlib/public/core/UInt128.swift:
```diff
@@ -280,7 +280,9 @@ extension UInt128 {
+    if _slowPath(other == .zero) {
+      return (self, true)
+    }
     // Unsigned divide never overflows.
```
This comment seems ironic now, doesn't it?

oscbyspro replied:
In a way, I suppose. I would say that the ironic part is the overflow return label, however, since it represents both proper overflow and another kind of failure (division by zero). Still, I can remove the comment if you find it unfitting. Generally, I don't mess with comments I did not write.

stephentyrone replied:
It's fine to leave it, I think.

oscbyspro committed:
Add a comment about signed remainder overflow semantics as requested.

I have paraphrased @stephentyrone's and @xwu's review comments to the best of my ability.

stephentyrone commented:
@swift-ci test

stephentyrone commented:
@oscbyspro do you have bandwidth to make another PR for the 6.1 branch? If not, I can do so for you.

oscbyspro commented:
I can do it with some patience on your part.

I'm currently reading the documentation for how Swift handles release branches:

- https://github.com/swiftlang/swift/blob/main/CONTRIBUTING.md
- https://github.com/swiftlang/swift/blob/main/docs/Branches.md

stephentyrone commented:
Why don't you give it a go, and if you get stuck I'll help.

Looking at an example is probably instructive: https://github.com/swiftlang/swift/pull/77699

oscbyspro commented:
I don't really `git cherry-pick` in my projects so I got a bit confused when the cherry-picked hashes were different. I suppose I thought `git` would combine the changes of each commit in isolation. In any case, I appreciate your streamlined instructions.
